presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -586,4 +592,163 @@ public void prepareFinal()
+        }
+
+        @Override
+        public void evaluateIntermediate(int groupId, BlockBuilder output)
+        {
+            if (blockBuilders == null) {
+                if (rawInputs == null) {
+                    return;
+                }
+                blockBuilders = new ObjectBigArray<>();  // we need to spill now; let's build the hash table here
+                for (GroupIdPage groupIdPage : rawInputs) {
+                    Page page = groupIdPage.getPage();
+                    GroupByIdBlock groupIdsBlock = groupIdPage.getGroupByIdBlock();
+                    for (int position = 0; position < page.getPositionCount(); position++) {
+                        long currentGroupId = groupIdsBlock.getGroupId(position);
+                        blockBuilders.ensureCapacity(currentGroupId);
+                        RowBlockBuilder rowBlockBuilder = blockBuilders.get(currentGroupId);
+                        if (rowBlockBuilder == null) {
+                            rowBlockBuilder = new RowBlockBuilder(spillingTypes, null, (int) groupIdsBlock.getGroupCount());
+                        }
+
+                        BlockBuilder currentOutput = rowBlockBuilder.beginBlockEntry();
+                        for (int channel = 0; channel < spillingTypes.size(); channel++) {
+                            spillingTypes.get(channel).appendTo(page.getBlock(channel), position, currentOutput);
+                        }
+                        rowBlockBuilder.closeEntry();
+
+                        blockBuilders.set(currentGroupId, rowBlockBuilder);
+                    }
+                }
+                rawInputs = null;
+                rawInputsSize = 0;
+            }
+
+            BlockBuilder singleArrayBlockWriter = output.beginBlockEntry();
+            checkState(rawInputs == null && blockBuilders != null);
+            checkState(singleArrayBlockWriter instanceof SingleArrayBlockWriter);
+
+            RowBlock rowBlock = (RowBlock) blockBuilders.get((long) groupId).build();
+            for (int i = 0; i < rowBlock.getPositionCount(); i++) {
+                singleArrayBlockWriter.appendStructure(rowBlock.getBlock(i));
+            }
+            output.closeEntry();
+        }
+
+        @Override
+        public void evaluateFinal(int groupId, BlockBuilder output)
+        {
+            // checkState(rawInputs == null && blockBuilders == null);
+            delegate.evaluateFinal(groupId, output);
+        }
+
+        @Override
+        public void prepareFinal()
+        {
+            checkState(rawInputs != null && blockBuilders == null);
+            for (GroupIdPage groupIdPage : rawInputs) {
+                delegate.addInput(groupIdPage.getGroupByIdBlock(), groupIdPage.getPage());
+            }
+
+            // cannot set raw inputs to null because evaluateIntermediate may be called after prepareFinal()
+            delegate.prepareFinal();
+        }
+
+        private static class GroupIdPage
```

Comment:
Made a pair class since otherwise we will have to separate group ids block and page every time we addInput to the delegate and make a new page object. Felt this was more efficient due to that.

Reply:
That is ok